File: src/benchmark.py
import time

def generate_random_strings(count=10000):
    # Dictionary to store strings and their counts
    string_counts = {}
    
    # Get current timestamp for initial seeds
    timestamp = int(time.time() * 1000)
    
    # Set up Xorshift+ state (needs two 64-bit seeds)
    state0 = timestamp ^ 0xDEADBEEF  # XOR with arbitrary constant
    state1 = (timestamp << 21) ^ 0x95419C24A637B12F  # Another arbitrary constant
    
    # Generate 10000 random strings
    for i in range(count):
        # Generate a pseudorandom string of length 8
        random_string = ""
        
        for _ in range(8):
            # Xorshift+ algorithm (much better statistical properties than LCG)
            s1 = state0 & 0xFFFFFFFFFFFFFFFF  # Ensure 64-bit unsigned
            s0 = state1
            
            # Update state
            state0 = s0
            s1 ^= s1 << 23
            s1 &= 0xFFFFFFFFFFFFFFFF  # Ensure 64-bit unsigned
            state1 = s1 ^ s0 ^ (s1 >> 18) ^ (s0 >> 5)
            state1 &= 0xFFFFFFFFFFFFFFFF  # Ensure 64-bit unsigned
            
            # Get random value
            rand_value = (state1 + s0) & 0xFFFFFFFFFFFFFFFF
            
            # Use just enough bits for our character range (0-61)
            rand_num = rand_value % 62
            
            # Convert to character
            if rand_num < 10:  # 0-9
                char = chr(ord('0') + rand_num)
            elif rand_num < 36:  # a-z
                char = chr(ord('a') + rand_num - 10)
            else:  # A-Z
                char = chr(ord('A') + rand_num - 36)
            
            random_string += char
        
        # Add to dictionary or increment count if it exists
        if random_string in string_counts:
            string_counts[random_string] += 1
        else:
            string_counts[random_string] = 1
    
    return string_counts


# Run the function 3 times and measure execution time
total_time = 0
for i in range(3):
    start_time = time.time()
    string_counts = generate_random_strings(10000)  # Generate 10000 strings
    end_time = time.time()
    
    # Convert seconds to milliseconds
    execution_time_ms = (end_time - start_time) * 1000
    total_time += execution_time_ms
    
    # Print in CSV format to console
    print(f"{execution_time_ms:.2f}\t",end="")

# Calculate and print the average execution time in milliseconds
average_time_ms = total_time / 3
print(f"{average_time_ms:.2f}")

File: src/benchmark.cpp
//
// Created by Alexander Pototskiy on 18.03.25.
//

#include <iostream>
#include <unordered_map>
#include <cstring>

#define STRING_COUNT 10000
#define STRING_LENGTH 8
#define HASH_TABLE_SIZE (2 * STRING_COUNT)  // Arbitrary size for hash table

uint64_t state0, state1;

template <size_t SizeV>
class small_fixed_string
{
    char data_[SizeV + 1];

public:
    inline small_fixed_string() noexcept
    {
        data_[SizeV] = 0;
    }

    inline char * data() noexcept { return data_; }
    inline char const* data() const noexcept { return data_; }

    friend inline bool operator==(small_fixed_string const& rls, small_fixed_string const& rhs) noexcept
    {
        return 0 == std::memcmp(rls.data_, rhs.data_, SizeV);
    }

    friend inline size_t hash_value(small_fixed_string const& s) noexcept
    {
        return std::hash<std::string_view>()(std::string_view{s.data_, SizeV});
    }
};

struct small_fixed_string_hash
{
    template <size_t SizeV>
    inline size_t operator()(small_fixed_string<SizeV> const& value) const noexcept { return hash_value(value); }
};

using string_type = small_fixed_string<STRING_LENGTH>;

std::unordered_map<string_type, uint64_t, small_fixed_string_hash> hashTable;

inline uint64_t xorshift_plus() noexcept {
    uint64_t s1 = state0;
    uint64_t s0 = state1;
    state0 = s0;
    s1 ^= s1 << 23;
    s1 ^= s1 >> 18;
    s1 ^= s0;
    s1 ^= s0 >> 5;
    state1 = s1;
    return state1 + s0;
}

static const char CHARSET[] = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

// Function to generate `count` random strings and store them in a hash table
void generate_random_strings(int count) {
    hashTable.clear();
    small_fixed_string<STRING_LENGTH> random_string;

    for (int i = 0; i < count; i++) {
        for (char * data = random_string.data(), *edata = data + STRING_LENGTH; data != edata; ++data) {
            *data = CHARSET[xorshift_plus() % 62];
        }

        if (auto it = hashTable.find(random_string); it == hashTable.end()) {
            hashTable.emplace_hint(it, random_string, 1);
        } else {
            it->second++;
        }
    }
}

double get_execution_time() {
    clock_t start, end;
    start = clock();
    generate_random_strings(STRING_COUNT);
    end = clock();
    return ((double)(end - start) / CLOCKS_PER_SEC) * 1000.0; // Convert to milliseconds
}

int main() {
    // Initialize random seed
    uint64_t timestamp = (uint64_t)time(NULL) * 1000;
    state0 = timestamp ^ 0xDEADBEEF; // Arbitrary constant
    state1 = (timestamp << 21) ^ 0x95419C24A637B12F; // Arbitrary constant

    double total_time = 0.0;
    hashTable.reserve(HASH_TABLE_SIZE);
    // Run 3 times and measure execution time
    for (int i = 0; i < 3; i++) {
        double exec_time = get_execution_time();
        total_time += exec_time;
        printf("%.2f\t", exec_time);
    }

    // Print average execution time
    printf("%.2f\n", total_time / 3);

    return 0;
}
